src/firefox/index.js

```diff
@@ -249,9 +281,45 @@ export async function useProfile(
     app,
     configureThisProfile = configureProfile,
     customPrefs = {},
+    createProfileFinder = defaultCreateProfileFinder,
   }: UseProfileParams = {},
 ): Promise<FirefoxProfile> {
-  const profile = new FirefoxProfile({destinationDirectory: profilePath});
+  let profile;
+  let defaultProfilePath = '';
+  let defaultDevProfilePath = '';
+  const finder = createProfileFinder();
+  try {
+    const dirExists = await isDirectory(profilePath);
+    if (await finder.hasProfileName('default')) {
+      defaultProfilePath = await finder.getPath('default');
+    }
+    if (await finder.hasProfileName('dev-edition-default')) {
+      defaultDevProfilePath = await finder.getPath('dev-edition-default');
+    }
+    if (dirExists) {
+      log.debug(`Using profile directory from "${profilePath}"`);
+      if (profilePath === defaultProfilePath ||
+        profilePath === defaultDevProfilePath) {
+        throw new WebExtError(
+          `Cannot use profile at "${profilePath}"`
+        );
+      }
+      profile = new FirefoxProfile({destinationDirectory: profilePath});
+    } else {
+      log.debug(`Assuming ${profilePath} is a named profile`);
+      if (profilePath === 'default' ||
+            profilePath === 'dev-edition-default') {
+        throw new WebExtError(
+          `Cannot use the blacklisted named profile "${profilePath}"`
+        );
+      }
+      const profileDirectory = await finder.getPath(profilePath);
+      profile = new FirefoxProfile({destinationDirectory: profileDirectory});
+    }
+  } catch (error) {
```

Comment:
I'm not sure that this giant `try...catch` is actually useful, at least by looking currently we can just leave any other exception to be propagated to the callers.